modules/arkit/arkit_interface.mm

```diff
@@ -0,0 +1,711 @@
+	new_tracker->set_type(ARVRServer::TRACKER_ANCHOR);
+
+	char tracker_name[256];
+	sprintf(tracker_name, "Anchor %02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x", p_uuid[0], p_uuid[1], p_uuid[2], p_uuid[3], p_uuid[4], p_uuid[5], p_uuid[6], p_uuid[7], p_uuid[8], p_uuid[9], p_uuid[10], p_uuid[11], p_uuid[12], p_uuid[13], p_uuid[14], p_uuid[15]);
+
+	String name = tracker_name;
+	print_line("Adding tracker " + name);
+	new_tracker->set_name(name);
+
+	// add our tracker
+	ARVRServer::get_singleton()->add_tracker(new_tracker);
+	anchors[num_anchors].tracker = new_tracker;
+	memcpy(anchors[num_anchors].uuid, p_uuid, 16);
+	num_anchors++;
+
+	return new_tracker;
+}
+
+void ARKitInterface::remove_anchor_for_uuid(const unsigned char *p_uuid) {
+	if (anchors == NULL) {
+		// ignore
+	} else {
+		for (unsigned int i = 0; i < num_anchors; i++) {
+			if (memcmp(anchors[i].uuid, p_uuid, 16) == 0) {
+				// remove our tracker
+				ARVRServer::get_singleton()->remove_tracker(anchors[i].tracker);
+				memdelete(anchors[i].tracker);
+
+				// bring remaining forward
+				for (unsigned int j = i + 1; j < num_anchors; j++) {
+					anchors[j - 1] = anchors[j];
+				};
+
+				// decrease count
+				num_anchors--;
+				return;
+			}
+		}
+	}
+}
+
+void ARKitInterface::remove_all_anchors() {
+	if (anchors != NULL) {
+		for (unsigned int i = 0; i < num_anchors; i++) {
+			// remove our tracker
+			ARVRServer::get_singleton()->remove_tracker(anchors[i].tracker);
+			memdelete(anchors[i].tracker);
+		};
+
+		free(anchors);
+		anchors = NULL;
+		num_anchors = 0;
+	}
+}
+
+void ARKitInterface::process() {
+	_THREAD_SAFE_METHOD_
+
+	if (@available(iOS 11.0, *)) {
+		if (initialized) {
+			// get our next ARFrame
+			ARFrame *current_frame = ar_session.currentFrame;
+			if (last_timestamp != current_frame.timestamp) {
+				// only process if we have a new frame
+				last_timestamp = current_frame.timestamp;
+
+				// get some info about our screen and orientation
+				Size2 screen_size = OS::get_singleton()->get_window_size();
+				UIDeviceOrientation orientation = [[UIApplication sharedApplication] statusBarOrientation];
+
+				CameraFeed *feed = CameraIOS::get_arkit_feed();
+
+				// Grab our camera image for our backbuffer
+				CVPixelBufferRef pixelBuffer = current_frame.capturedImage;
+				if ((CVPixelBufferGetPlaneCount(pixelBuffer) == 2) && (feed != NULL)) {
+					// Plane 0 is our Y and Plane 1 is our CbCr buffer
+
+					// ignored, we check each plane separately
+					// image_width = CVPixelBufferGetWidth(pixelBuffer);
+					// image_height = CVPixelBufferGetHeight(pixelBuffer);
+
+					//				printf("Pixel buffer %i - %i\n", image_width, image_height);
+
+					// It says that we need to lock this on the documentation pages but it's not in the samples
+					// need to lock our base address so we can access our pixel buffers, better safe then sorry?
+					CVPixelBufferLockBaseAddress(pixelBuffer, kCVPixelBufferLock_ReadOnly);
```

Comment:
I believe a lock is always necessary if you're moving from GPU to CPU with a `CVPixelBufferRef`

Reply:
Removed the comments, its safer locking it any way you look at it :)